File: android/DKParkingSDK/src/main/kotlin/com/dkparking/sdk/core/ParkingEvaluationResult.kt
// ParkingEvaluationResult.kt
// DK Parking Engine SDK — Android Version 1 (Phase 9 Android)
// Implements the full output contract per OUTPUT_CONTRACT.md
// Parity: iOS ParkingEvaluationResult.swift (PC-005, PC-007)

package com.dkparking.sdk.core

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.util.UUID

// MARK: - Top-level result

/**
 * The complete structured output from a single parking evaluation call.
 * All fields are normative per OUTPUT_CONTRACT.md.
 */
@Serializable
data class ParkingEvaluationResult(
    val evaluationId: String = UUID.randomUUID().toString(),
    val timestampEpochMs: Long = System.currentTimeMillis(),
    val decisionState: DecisionState,
    val refusalReasons: List<RefusalReasonCode> = emptyList(),
    /** Top-level rule family being evaluated (REQUIRED per OUTPUT_CONTRACT.md §2). */
    val ruleFamily: RuleFamily,
    val measurement: MeasurementBundle? = null,
    val targetInfo: TargetInfo? = null,
    val featureCandidate: FeatureCandidateInfo? = null,
    val captureQuality: CaptureQualityBundle,
    /** Advisory-first family outputs (empty list if none; REQUIRED per OUTPUT_CONTRACT.md §2). */
    val advisoryOutputs: List<AdvisoryOutput> = emptyList(),
    val unsupportedVisibleRestrictionFlag: Boolean = false,
    /** Locked limitations notice text (REQUIRED per OUTPUT_CONTRACT.md §2). */
    val limitationsNotice: String = "This app evaluates only specific supported Danish stopping and parking rules. Other rules, signs, and restrictions may apply. This is not legal advice.",
    val versionRefs: VersionRefs
)

// MARK: - MeasurementBundle

/**
 * Metric measurement details. Present only when a measurement was successfully computed.
 * Per OUTPUT_CONTRACT.md section 4.
 */
@Serializable
data class MeasurementBundle(
    val ruleFamily: String,
    val legalThresholdM: Double,
    val measuredDistanceM: Double,
    val signedMarginM: Double,
    val totalEstimatedErrorM: Double,
    val confidenceScore: Double,
    val measurementReferenceType: String,
    val boundaryProvenance: BoundaryProvenance,
    /** Which vehicle edge was used as the measurement reference (REQUIRED per OUTPUT_CONTRACT.md §4). */
    val vehicleEdgeUsed: String = "nearest_edge",
    val inNearThresholdZone: Boolean
) {
    companion object {
        fun create(
            ruleFamily: String,
            legalThresholdM: Double,
            measuredDistanceM: Double,
            signedMarginM: Double,
            totalEstimatedErrorM: Double,
            confidenceScore: Double,
            measurementReferenceType: String,
            boundaryProvenance: BoundaryProvenance,
            vehicleEdgeUsed: String = "nearest_edge",
            nearThresholdDowngradeMarginM: Double = 0.30
        ): MeasurementBundle = MeasurementBundle(
            ruleFamily = ruleFamily,
            legalThresholdM = legalThresholdM,
            measuredDistanceM = measuredDistanceM,
            signedMarginM = signedMarginM,
            totalEstimatedErrorM = totalEstimatedErrorM,
            confidenceScore = confidenceScore,
            measurementReferenceType = measurementReferenceType,
            boundaryProvenance = boundaryProvenance,
            vehicleEdgeUsed = vehicleEdgeUsed,
            inNearThresholdZone = kotlin.math.abs(signedMarginM) < (totalEstimatedErrorM + nearThresholdDowngradeMarginM)
        )
    }
}

// MARK: - BoundaryProvenance

/**
 * The provenance tier of the legal boundary used in measurement.
 * Locked values per legal_boundary_localization_strategy.md section 5.
 */
@Serializable
enum class BoundaryProvenance(val rawValue: String) {
    @SerialName("visual_detection")
    VISUAL_DETECTION("visual_detection"),

    @SerialName("map_prior_assisted")
    MAP_PRIOR_ASSISTED("map_prior_assisted"),

    @SerialName("map_prior_only")
    MAP_PRIOR_ONLY("map_prior_only")
}

// MARK: - TargetInfo

/**
 * Information about the confirmed target vehicle.
 * Per OUTPUT_CONTRACT.md section 5.
 */
@Serializable
data class TargetInfo(
    /** Identifier of the confirmed active target (REQUIRED per OUTPUT_CONTRACT.md §5). */
    val targetId: String = UUID.randomUUID().toString(),
    val targetConfirmationSource: TargetConfirmationSource,
    val footprintQualityScore: Double,
    val partialOcclusionDetected: Boolean
)

@Serializable
enum class TargetConfirmationSource(val rawValue: String) {
    @SerialName("auto_selected_unambiguous")
    AUTO_SELECTED_UNAMBIGUOUS("auto_selected_unambiguous"),

    @SerialName("user_confirmed")
    USER_CONFIRMED("user_confirmed")
}

// MARK: - FeatureCandidateInfo

/**
 * Information about the matched dataset feature candidate.
 * Per OUTPUT_CONTRACT.md section 6.
 */
@Serializable
data class FeatureCandidateInfo(
    val candidateFeatureId: String,
    val candidateFeatureType: String,
    val candidateSelectionBasis: String,
    val alternativeCandidatesRejected: Int,
    val candidateConfidenceScore: Double
)

// MARK: - CaptureQualityBundle

/**
 * AR session quality indicators at time of capture.
 * Per OUTPUT_CONTRACT.md section 8.
 */
@Serializable
data class CaptureQualityBundle(
    /** Camera focus quality score (REQUIRED per OUTPUT_CONTRACT.md §8). */
    val focusScore: Double = 1.0,
    /** Scene brightness quality score (REQUIRED per OUTPUT_CONTRACT.md §8). */
    val brightnessScore: Double = 1.0,
    val arPlaneStabilityScore: Double,
    val arMetricScaleValid: Boolean,
    val arMetricScaleScore: Double
)

// MARK: - AdvisoryOutput

/**
 * Advisory-first family output.
 * Per OUTPUT_CONTRACT.md section 9.
 */
@Serializable
data class AdvisoryOutput(
    /** The advisory family name (e.g., "driveway_obstruction"). */
    val advisoryFamily: String,
    /** "ADVISORY_CONCERN" or "ADVISORY_NO_CONCERN_DETECTED". */
    val advisoryState: String,
    /** Display label — MUST include the word "advisory". */
    val advisoryLabel: String,
    /** Optional additional context. */
    val advisoryNotes: String? = null
)

// MARK: - VersionRefs

/**
 * Version references for full traceability per VERSIONING_POLICY.md.
 * All fields are mandatory in every result.
 */
@Serializable
data class VersionRefs(
    val sdkVersion: String,
    val policyVersion: String,
    val datasetVersion: String,
    val datasetRegionId: String,
    val modelVersion: String,
    val legalSourceBaselineDate: String
)

File: ios/DKParkingSDK/Sources/DKParkingSDK/Core/ParkingEvaluationResult.swift
// ParkingEvaluationResult.swift
// DK Parking Engine SDK — Phase 9 Vertical Slice
// Implements the full output contract per OUTPUT_CONTRACT.md

import Foundation

// MARK: - Top-level result

/// The complete structured output from a single parking evaluation call.
/// All fields are normative per OUTPUT_CONTRACT.md.
public struct ParkingEvaluationResult: Codable {
    public let evaluationId: String
    public let timestamp: Date
    public let decisionState: DecisionState
    public let refusalReasons: [RefusalReasonCode]
    /// Top-level rule family being evaluated (REQUIRED per OUTPUT_CONTRACT.md §2).
    public let ruleFamily: RuleFamily
    public let measurement: MeasurementBundle?
    public let targetInfo: TargetInfo?
    public let featureCandidate: FeatureCandidateInfo?
    public let captureQuality: CaptureQualityBundle
    /// Advisory-first family outputs (empty list if none; REQUIRED per OUTPUT_CONTRACT.md §2).
    public let advisoryOutputs: [AdvisoryOutput]
    public let unsupportedVisibleRestrictionFlag: Bool
    /// Locked limitations notice text (REQUIRED per OUTPUT_CONTRACT.md §2, source: user_disclosures_and_copy.md §6).
    public let limitationsNotice: String
    public let versionRefs: VersionRefs

    public init(
        evaluationId: String = UUID().uuidString,
        timestamp: Date = Date(),
        decisionState: DecisionState,
        refusalReasons: [RefusalReasonCode] = [],
        ruleFamily: RuleFamily,
        measurement: MeasurementBundle? = nil,
        targetInfo: TargetInfo? = nil,
        featureCandidate: FeatureCandidateInfo? = nil,
        captureQuality: CaptureQualityBundle,
        advisoryOutputs: [AdvisoryOutput] = [],
        unsupportedVisibleRestrictionFlag: Bool = false,
        limitationsNotice: String = "This app evaluates only specific supported Danish stopping and parking rules. Other rules, signs, and restrictions may apply. This is not legal advice.",
        versionRefs: VersionRefs
    ) {
        self.evaluationId = evaluationId
        self.timestamp = timestamp
        self.decisionState = decisionState
        self.refusalReasons = refusalReasons
        self.ruleFamily = ruleFamily
        self.measurement = measurement
        self.targetInfo = targetInfo
        self.featureCandidate = featureCandidate
        self.captureQuality = captureQuality
        self.advisoryOutputs = advisoryOutputs
        self.unsupportedVisibleRestrictionFlag = unsupportedVisibleRestrictionFlag
        self.limitationsNotice = limitationsNotice
        self.versionRefs = versionRefs
    }
}

// MARK: - MeasurementBundle

/// Metric measurement details. Present only when a measurement was successfully computed.
/// Per OUTPUT_CONTRACT.md section 4.
public struct MeasurementBundle: Codable {
    public let ruleFamily: String
    public let legalThresholdM: Double
    public let measuredDistanceM: Double
    public let signedMarginM: Double
    public let totalEstimatedErrorM: Double
    public let confidenceScore: Double
    public let measurementReferenceType: String
    public let boundaryProvenance: BoundaryProvenance
    /// Which vehicle edge was used as the measurement reference (REQUIRED per OUTPUT_CONTRACT.md §4).
    public let vehicleEdgeUsed: String
    public let inNearThresholdZone: Bool

    public init(
        ruleFamily: String,
        legalThresholdM: Double,
        measuredDistanceM: Double,
        signedMarginM: Double,
        totalEstimatedErrorM: Double,
        confidenceScore: Double,
        measurementReferenceType: String,
        boundaryProvenance: BoundaryProvenance,
        vehicleEdgeUsed: String = "nearest_edge",
        nearThresholdDowngradeMarginM: Double = 0.30
    ) {
        self.ruleFamily = ruleFamily
        self.legalThresholdM = legalThresholdM
        self.measuredDistanceM = measuredDistanceM
        self.signedMarginM = signedMarginM
        self.totalEstimatedErrorM = totalEstimatedErrorM
        self.confidenceScore = confidenceScore
        self.measurementReferenceType = measurementReferenceType
        self.boundaryProvenance = boundaryProvenance
        self.vehicleEdgeUsed = vehicleEdgeUsed
        self.inNearThresholdZone = abs(signedMarginM) < (totalEstimatedErrorM + nearThresholdDowngradeMarginM)
    }
}

// MARK: - BoundaryProvenance

/// The provenance tier of the legal boundary used in measurement.
/// Locked values per legal_boundary_localization_strategy.md section 5.
public enum BoundaryProvenance: String, Codable {
    case visualDetection = "visual_detection"
    case mapPriorAssisted = "map_prior_assisted"
    case mapPriorOnly = "map_prior_only"
}

// MARK: - TargetInfo

/// Information about the confirmed target vehicle.
/// Per OUTPUT_CONTRACT.md section 5.
public struct TargetInfo: Codable {
    /// Identifier of the confirmed active target (REQUIRED per OUTPUT_CONTRACT.md §5).
    public let targetId: String
    public let targetConfirmationSource: TargetConfirmationSource
    public let footprintQualityScore: Double
    public let partialOcclusionDetected: Bool

    public init(
        targetId: String = UUID().uuidString,
        targetConfirmationSource: TargetConfirmationSource,
        footprintQualityScore: Double,
        partialOcclusionDetected: Bool
    ) {
        self.targetId = targetId
        self.targetConfirmationSource = targetConfirmationSource
        self.footprintQualityScore = footprintQualityScore
        self.partialOcclusionDetected = partialOcclusionDetected
    }
}

public enum TargetConfirmationSource: String, Codable {
    case autoSelectedUnambiguous = "auto_selected_unambiguous"
    case userConfirmed = "user_confirmed"
}

// MARK: - FeatureCandidateInfo

/// Information about the matched dataset feature candidate.
/// Per OUTPUT_CONTRACT.md section 6.
public struct FeatureCandidateInfo: Codable {
    public let candidateFeatureId: String
    public let candidateFeatureType: String
    public let candidateSelectionBasis: String
    public let alternativeCandidatesRejected: Int
    public let candidateConfidenceScore: Double

    public init(
        candidateFeatureId: String,
        candidateFeatureType: String,
        candidateSelectionBasis: String,
        alternativeCandidatesRejected: Int,
        candidateConfidenceScore: Double
    ) {
        self.candidateFeatureId = candidateFeatureId
        self.candidateFeatureType = candidateFeatureType
        self.candidateSelectionBasis = candidateSelectionBasis
        self.alternativeCandidatesRejected = alternativeCandidatesRejected
        self.candidateConfidenceScore = candidateConfidenceScore
    }
}

// MARK: - CaptureQualityBundle

/// AR session quality indicators at time of capture.
/// Per OUTPUT_CONTRACT.md section 8.
public struct CaptureQualityBundle: Codable {
    /// Camera focus quality score (REQUIRED per OUTPUT_CONTRACT.md §8).
    public let focusScore: Double
    /// Scene brightness quality score (REQUIRED per OUTPUT_CONTRACT.md §8).
    public let brightnessScore: Double
    public let arPlaneStabilityScore: Double
    public let arMetricScaleValid: Bool
    public let arMetricScaleScore: Double

    public init(
        focusScore: Double = 1.0,
        brightnessScore: Double = 1.0,
        arPlaneStabilityScore: Double,
        arMetricScaleValid: Bool,
        arMetricScaleScore: Double
    ) {
        self.focusScore = focusScore
        self.brightnessScore = brightnessScore
        self.arPlaneStabilityScore = arPlaneStabilityScore
        self.arMetricScaleValid = arMetricScaleValid
        self.arMetricScaleScore = arMetricScaleScore
    }
}

// MARK: - AdvisoryOutput

/// Advisory-first family output.
/// Per OUTPUT_CONTRACT.md section 9.
public struct AdvisoryOutput: Codable {
    /// The advisory family name (e.g., `driveway_obstruction`).
    public let advisoryFamily: String
    /// `ADVISORY_CONCERN` or `ADVISORY_NO_CONCERN_DETECTED`.
    public let advisoryState: String
    /// Display label — MUST include the word "advisory".
    public let advisoryLabel: String
    /// Optional additional context.
    public let advisoryNotes: String?

    public init(
        advisoryFamily: String,
        advisoryState: String,
        advisoryLabel: String,
        advisoryNotes: String? = nil
    ) {
        self.advisoryFamily = advisoryFamily
        self.advisoryState = advisoryState
        self.advisoryLabel = advisoryLabel
        self.advisoryNotes = advisoryNotes
    }
}

// MARK: - VersionRefs

/// Version references for full traceability per VERSIONING_POLICY.md.
/// All fields are mandatory in every result.
public struct VersionRefs: Codable {
    public let sdkVersion: String
    public let policyVersion: String
    public let datasetVersion: String
    public let datasetRegionId: String
    public let modelVersion: String
    public let legalSourceBaselineDate: String

    public init(
        sdkVersion: String,
        policyVersion: String,
        datasetVersion: String,
        datasetRegionId: String,
        modelVersion: String,
        legalSourceBaselineDate: String
    ) {
        self.sdkVersion = sdkVersion
        self.policyVersion = policyVersion
        self.datasetVersion = datasetVersion
        self.datasetRegionId = datasetRegionId
        self.modelVersion = modelVersion
        self.legalSourceBaselineDate = legalSourceBaselineDate
    }
}
